rasa_core/training/structures.py

```diff
@@ -89,36 +111,150 @@ def create_copy(self, use_new_id):
+        )
+        self.as_story_string_helper.form_string += self._bot_string(
+                story_step_element, prefix
+        )
+
+    def _reset_stored_strings(self):
+        self.as_story_string_helper.form_string = ''
+        self.as_story_string_helper.no_form_string = ''
 
     def as_story_string(self, flat=False, e2e=False):
         # if the result should be flattened, we
         # will exclude the caption and any checkpoints.
+
+        for s in self.start_checkpoints:
+            if s.name == STORY_START:
+                # first story step in the story, so reset helper
+                self.as_story_string_helper = AsStoryStringHelper()
+
         if flat:
             result = ""
         else:
             result = "\n## {}\n".format(self.block_name)
             for s in self.start_checkpoints:
                 if s.name != STORY_START:
-                    result += "> {}\n".format(s.as_story_string())
+                    result += self._checkpoint_string(s)
+
         for s in self.events:
             if isinstance(s, UserUttered):
-                result += "* {}\n".format(s.as_story_string(e2e))
+                if self.as_story_string_helper.active_form is None:
+                    result += self._user_string(s, e2e)
+                else:
+                    # form is active
+                    # it is not known whether the form will be
+                    # successfully executed, so store this
+                    # story string for later
+                    self._store_user_strings(s, e2e, FORM_PREFIX)
+
+            elif isinstance(s, Form):
+                # form got either activated or deactivated
+                self.as_story_string_helper.active_form = s.name
+
+                if self.as_story_string_helper.active_form is None:
+                    # form deactivated, so form succeeded,
+                    # so add story string with form prefix
+                    result += self.as_story_string_helper.form_string
+                    # remove all stored story strings
+                    self._reset_stored_strings()
+
+                result += self._bot_string(s)
+
+            elif isinstance(s, FormValidation):
+                self.as_story_string_helper.form_validation = s.validate
+
+            elif isinstance(s, ActionExecutionRejected):
+                if (s.action_name ==
+                        self.as_story_string_helper.active_form):
+                    # form rejected
+                    self.as_story_string_helper.form_rejected = True
+
+            elif isinstance(s, ActionExecuted):
+                if self._is_action_listen(s):
+                    pass
+                elif self.as_story_string_helper.active_form is None:
+                    result += self._bot_string(s)
+                else:
+                    # form is active
+                    if self.as_story_string_helper.form_rejected:
+                        if (self.as_story_string_helper.form_validation and
```

Comment:
can we change this `as_story_string_helper` to have a shorter name, reading this is giving me a headache

Reply:
I know, but I didn't come up with better name that is not confusing. Do you have an idea?

Reply:
can we even just rename it to `story_string_helper`? like even just removing one word helps i think

Reply:
👍